iree/turbine/kernel/wave/in_thread_transpose.py

```diff
@@ -0,0 +1,335 @@
+        load_elems_per_thread, expected_number_of_loads = (
+            expected_number_of_stores,
+            store_elems_per_thread,
+        )
+        load_shape = get_tiled_shape(
+            materialize_shape(constraint_tile_size, src_symbolic_shape),
+            load_elems_per_thread,
+            expected_number_of_loads,
+        )
+        logger.info(f"load_shape={load_shape}")
+
+        # Construct new reads.
+        for i in range(expected_number_of_loads):
+            read_index = get_tiled_index(
+                linear_id,
+                load_shape,
+                load_elems_per_thread,
+                expected_number_of_loads,
+                i,
+                transpose=False,
+            )
+            read_index = {
+                k: IndexSequence(v, 1, 1)
+                for k, v in zip(src_symbolic_shape, read_index)
+            }
+            read_index = combine_index(
+                global_index,
+                read_index,
+                src_symbolic_shape[-1],
+                load_elems_per_thread,
+            )
+            logger.info(f"read_index={read_index}")
+            mapping = read.mapping
+            if mapping is not None:
+                # As we are transposing read, update mapping to still be output
+                # identity.
+                out_mapping = {
+                    k: IndexMapping.iterator(i)
+                    for i, k in enumerate(src_symbolic_shape)
+                }
+                subs = {v: out_mapping[k] for k, v in mapping.output_mapping.items()}
+                input_mapping = {
+                    k: safe_subs(v, subs, simultaneous=True)
+                    for k, v in mapping.input_mapping.items()
+                }
+                mapping = IndexMapping(
+                    num_iterators=len(out_mapping),
+                    inputs=input_mapping,
+                    outputs=out_mapping,
+                    dynamic_val_mappings=mapping.dynamic_val_mappings,
+                )
+
+            with read.graph.inserting_before(read.fx_node):
+                new_read = Read(
+                    read.memory,
+                    load_elems_per_thread,
+                    mapping=mapping,
+                    mapping_dynamic_vals=read.mapping_dynamic_vals,
+                ).add_to_graph(read.graph)
+                new_read.index = read_index
+                new_read.vector_shapes = read.vector_shapes
+                new_read_custom = get_custom(new_read)
+                new_read_custom.infer_type()
+                update_read_mapping_dynamic_values(new_read_custom)
+                new_reads.append(new_read)
+
+        new_writes = defaultdict(list)
+        repacked = []
+
+        store_shape = get_tiled_shape(
+            materialized_shape, store_elems_per_thread, expected_number_of_stores
+        )
+        logger.info(f"store_shape={store_shape}")
+
+        # Construct transpose.
```

Comment:
How is Reshape able to figure it out? Maybe a comment with an example will help.

Reply:
when `Reshape` has multiple inputs it's just concatenates all of them into single vector